### PR TITLE
self extend support

### PR DESCRIPTION
**Pull request by cognitivetech**
trying to add support for self-extend as discussed here: https://github.com/ollama/ollama/issues/1964

I was hoping it would be as simple as adding these parameters as I've seen done in a previous commit, but I was copying moves from an older configuration of the source.

obviously I'm missing something.. probably rudimentary.

```
llm/dyn_ext_server.go:102:10: sparams.g_size undefined (type _Ctype_struct_ext_server_params has no field or method g_size)
llm/dyn_ext_server.go:103:10: sparams.w_size undefined (type _Ctype_struct_ext_server_params has no field or method w_size)
```

Hoping some more seasoned golang developer will help from here

### CODE REVIEW

**Comment by AliArmani3397** [comment]
@cognitivetech is the self-extend parameter set pass in the modelfile?

**Comment by cognitivetech** [comment]
I got the error just trying to build the app.. I'm really in over my head here.. will just try self-extend where its easier to pass parameters

**Comment by AliArmani3397** [comment]
@cognitivetech right, why building it though! this parameter seems to be well set in llama.cpp though it's not very functional at the moment until flash attention gets merged there. 

I'm new to ollama, but doesn't any new merge in Ollama's backend ( llama.cpp) automatically gets added to Ollama as well?

**Comment by cognitivetech** [comment]
> @cognitivetech right, why building it though! this parameter seems to be well set in llama.cpp though it's not very functional at the moment until flash attention gets merged there.
> 
> I'm new to ollama, but doesn't any new merge in Ollama's backend ( llama.cpp) automatically gets added to Ollama as well?

Each parameter must be added to ollama code. You can see in the files changed and previous PR how that has been done, all my research into this problem is documented in the related issue https://github.com/ollama/ollama/issues/1964

Buut.. the code changed since the original PR that i was basing my attempt on, and I did some guess-work to try and update accordingly

**Comment by AliArmani3397** [comment]
@cognitivetech alright. now it makes sense. not related to this PR, But saw you discussing Miqu system prompts in another ollama's PR, would be interested to know if you successfully implemented some sort of context extension technique to Miqu. There are some interesting papers along with their codes available here. Lucid is also cooking some recipe combining the advantages of a couple of techniques to make context extension more efficient (https://github.com/lucidrains/ring-attention-pytorch). but all require fine tuning and some even make the model power hungry like there is no tomorrow. for now i just want to test the non FT methods on Miqu as I dont feel comfortable FTing an already quant model. Anyways, would interested to know if you had any luck with Miqu or other 70B models context extension.

File: api/types.go [hunk 1]
```diff
@@ -125,6 +125,8 @@ type Runner struct {
 	RopeFrequencyBase  float32 `json:"rope_frequency_base,omitempty"`
 	RopeFrequencyScale float32 `json:"rope_frequency_scale,omitempty"`
 	NumThread          int     `json:"num_thread,omitempty"`
+	GSize              int     `json:"g_size,omitempty"`
```
Comment (Latrasis) [inline api/types.go:128]: @cognitivetech  Doesn't llama.cpp accept these params: `grp-attn-n` and `grp-attn-w` ? Not sure where you got `g_size` and `w_size` from.

```go
GSize int `json:"grp-attn-n,omitempty"`
WSize int `json:"grp-attn-w,omitempty"`
```
https://github.com/ggerganov/llama.cpp/issues/4886#issuecomment-1890465266

**Pull request closed without merging**